webhook_service/main.go

```diff
@@ -0,0 +1,873 @@
+		// webhooks from
+		rateLimit := &RateLimitState{}
+		for attempts := 0; attempts < maxMsgReattempts && (!msg.Discardable); attempts++ {
+			bucket, err = SendWebhook(rdb, &msg, rateLimit)
+			if err == nil {
+				rdb.HSet(ctx, bucketsHash, webhookID, bucket)
+				// Start worker if new bucket
+				if _, exists := GetWorker(bucket, mu); !exists {
+					// Make a new worker but use the response we got from Discord
+					// for the first time the message was sent
+					worker := NewBucketWorker(bucket, rdb)
+					worker.rateLimit.RateLimited = rateLimit.RateLimited
+					worker.rateLimit.Remaining = rateLimit.Remaining
+					worker.rateLimit.ResetAfter = rateLimit.ResetAfter
+					worker.rateLimit.ResetTime = rateLimit.ResetTime
+					SetWorker(bucket, worker, mu)
+					go worker.ProcessQueue(rdb)
+				}
+			} else {
+				logger.Error(fmt.Sprintf("First-time send error: %v", err))
+				if rateLimit.RateLimited {
+					wait := GetWorkerRateLimitSleepTime(rateLimit)
+					time.Sleep(wait)
+				} else {
+					// We had some sort of (non rate limiting error) trying to send this message
+					// back off before attempting again
+					time.Sleep(time.Second * 2 * time.Duration(attempts+1))
+				}
+			}
+		}
+		queueKey := bucketQueuePrefix + bucket
+		rdb.LTrim(ctx, queueKey, 0, maxQueueSize-1)
+	}
+}
+
+func ProcessTransient(rdb *redis.Client, msg Message, mu *sync.Mutex) {
+	webhookID, err := ExtractWebhookID(msg.Payload.URL)
+	if err != nil {
+		logger.Error(fmt.Sprintf("could not parse webhook ID from: %s:%s", msg.Payload.URL, err))
+		return
+	}
+
+	ctx := context.Background()
+	bucket, err := rdb.HGet(ctx, bucketsHash, webhookID).Result()
+	var worker *BucketWorker
+	if err != redis.Nil {
+		worker, _ = GetWorker(bucket, mu)
+	}
+
+	if worker == nil {
+		worker = NewBucketWorker(bucket, rdb)
+		SetWorker(bucket, worker, mu)
+	}
+
+	// No need to try reattempts; these messages are inherently discardable
+	bucket, err = SendWebhook(rdb, &msg, worker.rateLimit)
+	if err != nil {
+		logger.Error(fmt.Sprintf("transient message send error for %s: %v", msg.Payload.WebhookID, err))
+	}
+	rdb.HSet(ctx, bucketsHash, webhookID, bucket)
+}
+
+// Subscribe to transient messages via pub/sub
+func SubscribeTransients(rdb *redis.Client, mu *sync.Mutex) {
+	ctx := context.Background()
+	pubsub := rdb.Subscribe(ctx, transientChannel)
+	defer pubsub.Close()
+
+	logger.Info(fmt.Sprintf("Subscribing to: %s", transientChannel))
+	for msg := range pubsub.Channel() {
+		var transientMsg Message
+		if err := json.Unmarshal([]byte(msg.Payload), &transientMsg); err != nil {
+			logger.Error(fmt.Sprintf("transient unmarshal error: %v, JSON: %s", err, msg.Payload))
+			continue
+		}
+		go ProcessTransient(rdb, transientMsg, mu)
+	}
+}
+
+func main() {
+	logger.Info("Starting service")
+	rdb := SetupRedis()
+
+	// For the Docker health check
+	path := filepath.Join("/app", "webhook-service-healthy")
+	logger.Info(path)
+	exec.Command("touch", path).Start()
```

Comment:
shelling out to create a file... ooof :) What about using `os.Create` instead? :)